File: swarmmanagercontroller/src/main/java/com/swarmmanager/docker/api/common/AbstractFilters.java
package com.swarmmanager.docker.api.common;

import javax.json.Json;
import javax.json.JsonObjectBuilder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class AbstractFilters implements Filters {

    private Map<String, List<String>> filterParameters;

    public AbstractFilters() {
        filterParameters = new HashMap<>();
    }

    public String getJson() {
        JsonObjectBuilder jsonObjectBuilder = Json.createObjectBuilder();
        if (filterParameters != null) {
            for (Map.Entry<String, List<String>> filterEntry : filterParameters.entrySet()) {
                if (filterEntry.getValue() != null) {
                    JsonObjectBuilder valuesJsonBuilder = Json.createObjectBuilder();
                    for (String value : filterEntry.getValue()) {
                        valuesJsonBuilder.add(value, true);
                    }
                    jsonObjectBuilder.add(filterEntry.getKey(), valuesJsonBuilder);
                }
            }
        }
        return jsonObjectBuilder.build().toString();
    }

    protected AbstractFilters addFilter(String key, String value) {
        if (value != null) {
            List<String> values = filterParameters.computeIfAbsent(key, k -> new ArrayList<>());
            values.add(value);
        }
        return this;
    }

}

File: swarmmanagercontroller-new/src/main/kotlin/com/swarmbit/docker/api/common/AbstractFilters.kt
package com.swarmbit.docker.api.common

import com.swarmbit.extensions.urlEncode
import javax.json.Json
import kotlin.collections.component1
import kotlin.collections.component2

abstract class AbstractFilters : Filters {
    private val filterParameters: MutableMap<String, MutableList<Filter>> = mutableMapOf()

    override fun asUrlJson(): String {
        val jsonObjectBuilder = Json.createObjectBuilder()
        for ((key, values) in filterParameters) {
            val valuesJsonBuilder = Json.createObjectBuilder()
            for (value in values) {
                valuesJsonBuilder.add(value.value, value.equals)
            }
            jsonObjectBuilder.add(key, valuesJsonBuilder)
        }
        return jsonObjectBuilder.build().toString().urlEncode()
    }

    protected fun addFilter(key: String, value: String, equals: Boolean = true): AbstractFilters {
        val values = filterParameters.computeIfAbsent(key) { mutableListOf() }
        values.add(Filter(value, equals))
        return this
    }

}

private data class Filter(val value: String, val equals: Boolean)